File: python/src/anysearch/providers/jina.py
"""Jina Reader Search (s.jina.ai) — search that returns clean page content."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability


class JinaProvider(BaseProvider):
    name = "jina"
    env_keys = ("JINA_API_KEY",)
    default_base_url = "https://s.jina.ai"
    capabilities = frozenset({Capability.CONTENT})

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
        }
        # Ask for links only when full content was not requested (faster, cheaper).
        if not req.include_content:
            headers["X-Respond-With"] = "no-content"
        params = {"q": req.query}
        params.update(req.extra)
        return PreparedRequest(
            method="GET",
            url=f"{self.base_url}/",
            headers=headers,
            params=params,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        items = data.get("data") or []
        if isinstance(items, dict):
            items = [items]
        results = []
        for item in items:
            content = item.get("content")
            description = item.get("description")
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("url"),
                    snippet=description or (content[:300] if content else None),
                    text=content if req.include_content else None,
                    published_date=item.get("date") or item.get("publishedTime"),
                    raw=item,
                )
            )
        return self._response(req, results, data, elapsed_ms=elapsed_ms)

File: js/src/providers/jina.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

export const jina: ProviderSpec = {
  name: "jina",
  envKeys: ["JINA_API_KEY"],
  defaultBaseUrl: "https://s.jina.ai",
  capabilities: new Set(["content"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${ctx.apiKey ?? ""}`,
      Accept: "application/json",
    };
    if (!req.includeContent) headers["X-Respond-With"] = "no-content";
    const params: Record<string, string | number | boolean | undefined> = { q: req.query };
    for (const [k, v] of Object.entries(req.extra)) params[k] = v as string;
    return { method: "GET", url: `${ctx.baseUrl}/`, headers, params };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    let items = data.data ?? [];
    if (!Array.isArray(items)) items = [items];
    const results = items.map((item: any) => {
      const content: string | undefined = item.content;
      return result({
        title: item.title,
        url: item.url,
        snippet: item.description ?? (content ? content.slice(0, 300) : undefined),
        text: req.includeContent ? content : undefined,
        publishedDate: item.date ?? item.publishedTime,
        raw: item,
      });
    });
    return finalize(jina, req, results, data, { elapsedMs });
  },
};
